### PR TITLE
Feat/reward pool refactor #1976

### PR DESCRIPTION
**Pull request by shannonwells**
# Goal
The goal of this PR is to implement a "chunk" version of the overall reward pool history to reduce read/write load and hence weight for transactions and `on_initialize` when a new `RewardEra` needs to start.

Part of #1976

# Discussion
This is is just one part of optimizing storage access.  There will be more work later.  

The current era Reward Pool and a `ProviderBoostHistories` record are updated on every `provider_boost` and `unstake` (for a boost account) call.

Instead of storing all overall Reward Pool history in a single `CountedStorageMap`, store the current total stake for the current era separately, and divide the actual past era history up into `BoundedBTrees`. The BTrees are of size `RewardPoolChunkLength`.  These are in a StorageMap indexed by a u32 which is just the chunk index from 0 to history-size/chunk-length - 1.

This is unsurprisingly more complicated than the original solution, however it does have some advantages.

* It reduces reads and writes when there is a new Reward Era and for claiming and listing unclaimed rewards. Because each chunk is a BTree, on a new Reward Era you just remove the oldest and insert the newest into each chunk.
* It reduces reads when listing unclaimed rewards, and will reduce needed reads+writes when `claim_rewards` is implemented (because it will use the same function).
* It slightly reduces the weight of a stake/unstake because the current total stake is stored separately from Reward Pool History.
* It reduces the storage size of the reward pool history, generally, to just the total staked instead of a struct with 2 other values that aren't needed.

# Checklist
- [x] Tests added
- [x] Benchmarks added
- [x] Weights updated
- [x] Migration to initialize `ProviderBoostRewardPools` storage and `CurrentEraProviderBoostTotal`
- [ ] Verify migration works

### CODE REVIEW

**Review by wilwade — COMMENTED** [review]
Wanting to go ahead and dump ideas. Got through most of `pallets/capacity/src/lib.rs`

File: pallets/capacity/src/lib.rs [hunk 1]
```diff
@@ -703,34 +717,33 @@ impl<T: Config> Pallet<T> {
...
-		reward_pool.total_staked_token = reward_pool.total_staked_token.saturating_add(*amount);
-
-		Self::set_staking_account_and_lock(staker, staking_details)?;
-		Self::set_target_details_for(staker, *target, target_details);
-		Self::set_capacity_for(*target, capacity_details);
-		Self::set_reward_pool(era, &reward_pool);
 		Self::upsert_boost_history(staker, era, *amount, true)?;
+
+		let reward_pool_total = CurrentEraProviderBoostTotal::<T>::get();
+		CurrentEraProviderBoostTotal::<T>::set(reward_pool_total.saturating_add(*amount));
```
[the remark this answers is not in the document]
Comment (shannonwells) [reply pallets/capacity/src/lib.rs:745]: I have no objections, but we've been going for the get/set pattern in the last year or so. I'm following our selected pattern.

File: pallets/capacity/src/lib.rs [hunk 2]
```diff
@@ -1141,13 +1131,98 @@ impl<T: Config> Pallet<T> {
...
+		let reward_pool_chunk = Self::get_reward_pool_chunk(chunk_idx).unwrap_or_default(); // 1r
+		let total_for_era =
+			reward_pool_chunk.total_for_era(&reward_era).ok_or(Error::<T>::EraOutOfRange)?;
+		Ok(*total_for_era)
+	}
+
+	pub(crate) fn get_chunk_index_for_era(
+		era: T::RewardEra,
+		current_era: T::RewardEra,
+	) -> Option<u32> {
```
[the remark this answers is not in the document]
Comment (wilwade) [reply pallets/capacity/src/lib.rs:1172]: No it shouldn't.

File: pallets/capacity/src/lib.rs [hunk 3]
```diff
@@ -1141,13 +1131,98 @@ impl<T: Config> Pallet<T> {
...
+
+		let history_limit: u32 = T::ProviderBoostHistoryLimit::get();
+		let era_diff = current_era.saturating_sub(era);
+		if era_diff > history_limit.into() {
+			return None;
+		}
+
+		let chunk_len = T::RewardPoolChunkLength::get();
+		let chunks: u32 = history_limit.saturating_div(chunk_len);
+		(0u32..chunks).find(|&i| era_diff.le(&(chunk_len * (i + 1)).into()))
```
[the remark this answers is not in the document]
Comment (wilwade) [reply pallets/capacity/src/lib.rs:1185]: Notes, but going back to re-read some stuff.

- History Length = 6
- Chunk size = 2
- Chunks stored = (History Length / Chunk size) + 1
- Chunk Index = Floor((era % (History Length + chunk size)) / chunk size)

```
[1,2], [3,4], [5,6], [_, _]
[1,2], [3,4], [5,6], [7,_]
[1, 2], [3,4], [5,6], [7,8]
[9, _], [3,4], [5,6], [7,8]
[9, 10], [3,4], [5,6], [7,8]
[9, 10], [11, _], [5,6], [7,8]
```

**Review by wilwade — APPROVED** [review]
🚢 it

**Pull request merged**